File: observability/ts/bootstrap.ts
import { NodeSDK } from "@opentelemetry/sdk-node";
import { OTLPTraceExporter } from "@opentelemetry/exporter-trace-otlp-grpc";
import { Resource } from "@opentelemetry/resources";
import { SemanticResourceAttributes } from "@opentelemetry/semantic-conventions";


/**
 * Standard OTel wiring every Node service gets by depending on this target
 * (directly, or transitively via the `ts_service` macro).
 * Mirrors tools/observability/jvm/OtelBootstrap.kt.
 * */
export function initOtel(serviceName: string): NodeSDK {
	const sdk = new NodeSDK({
		resource: new Resource({
			[SemanticResourceAttributes.SERVICE_NAME]: serviceName,
		}),
		traceExporter: new OTLPTraceExporter({
			url: process.env.OTEL_EXPORTER_OTLP_ENPOINT ?? "http://otel-collector:4317",
		}),
	});

	sdk.start();
	return sdk;
}

File: observability/rust/bootstrap.rs
//! Standard OTel wiring every Rust service gets by depending on this
//! crate — mirrors bootstrap.go / OtelBootstrap.kt / otel-bootstrap.ts.

use opentelemetry::global;
use opentelemetry_otlp::WithExportConfig;
use opentelemetry_sdk::{trace as sdktrace, Resource};

pub fn init(service_name: &str) -> sdktrace::Tracer {
    let endpoint = std::env::var("OTEL_EXPORTER_OTLP_ENDPOINT")
        .unwrap_or_else(|_| "http://otel-collector:4317".to_string());

    let resource = Resource::new(vec![opentelemetry::KeyValue::new(
        "service.name",
        service_name.to_string(),
    )]);

    let tracer = opentelemetry_otlp::new_pipeline()
        .tracing()
        .with_exporter(
            opentelemetry_otlp::new_exporter()
                .tonic()
                .with_endpoint(endpoint),
        )
        .with_trace_config(sdktrace::config().with_resource(resource))
        .install_batch(opentelemetry_sdk::runtime::Tokio)
        .expect("failed to install OTel pipeline");

    global::set_text_map_propagator(opentelemetry_sdk::propagation::TraceContextPropagator::new());
    tracer
}